### PR TITLE
rpcclient: fix getblocknotifications no filter case handling

### PR DESCRIPTION
**Pull request by AliceInHunterland**

### CODE REVIEW

File: pkg/rpcclient/rpc.go [hunk 1]
```diff
@@ -974,9 +974,17 @@ func (c *Client) GetRawNotaryPool() (*result.RawNotaryPool, error) {
 }
 
 // GetBlockNotifications returns notifications from a block organized by trigger type.
-func (c *Client) GetBlockNotifications(blockHash util.Uint256, filters ...*neorpc.NotificationFilter) (*result.BlockNotifications, error) {
+func (c *Client) GetBlockNotifications(blockHash util.Uint256, filter ...*neorpc.NotificationFilter) (*result.BlockNotifications, error) {
```
Comment (AnnaShaleva) [inline pkg/rpcclient/rpc.go:977]: Why do we ever have filters specified via `...`? Change it to `filter *neorpc.NotificationFilter` to match WSClient's notation of `Receive*` functions. It's an incompatible change, but it's trivial to handle on the user's side.

@roman-khimov ACK?

**Review by AnnaShaleva — COMMENTED** [review]
It may be useful to be able to provide multiple filters with `||` meaning, just to allow user to use a single request instead of multiple precise requests (or instead of a single request that is too generic). So I'd suggest to:
1. Extend client to send multiple filters
2. Extend server to accept multiple filters
3. Document the `||` filter concatination behaviour.

But need @roman-khimov's ACK for that. And this way it will be a compatible change.

File: pkg/rpcclient/rpc.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/rpcclient/rpc.go:977]: ACK

**Comment by roman-khimov** [comment]
Keep it simple, we don't need `or` logic now.

**Review by roman-khimov — APPROVED** [review]

File: pkg/rpcclient/rpc.go [hunk 2]
```diff
@@ -974,9 +974,9 @@ func (c *Client) GetRawNotaryPool() (*result.RawNotaryPool, error) {
 }
 
 // GetBlockNotifications returns notifications from a block organized by trigger type.
-func (c *Client) GetBlockNotifications(blockHash util.Uint256, filters ...*neorpc.NotificationFilter) (*result.BlockNotifications, error) {
+func (c *Client) GetBlockNotifications(blockHash util.Uint256, filter *neorpc.NotificationFilter) (*result.BlockNotifications, error) {
 	var resp = &result.BlockNotifications{}
-	if err := c.performRequest("getblocknotifications", []any{blockHash.StringLE(), filters}, resp); err != nil {
+	if err := c.performRequest("getblocknotifications", []any{blockHash.StringLE(), filter}, resp); err != nil {
```
Comment (AnnaShaleva) [inline pkg/rpcclient/rpc.go:979]: You need to check for nil and then create a copy. See how `ReceiveExecutionNotifications` works.

Comment (AnnaShaleva) [reply pkg/rpcclient/rpc.go:979]: And add a test for nil case to `pkg/rpcclient/rpc_test.go`

File: pkg/rpcclient/rpc.go [hunk 3]
```diff
@@ -974,9 +974,15 @@ func (c *Client) GetRawNotaryPool() (*result.RawNotaryPool, error) {
...
-	if err := c.performRequest("getblocknotifications", []any{blockHash.StringLE(), filters}, resp); err != nil {
+func (c *Client) GetBlockNotifications(blockHash util.Uint256, filter *neorpc.NotificationFilter) (*result.BlockNotifications, error) {
+	var (
+		resp = &result.BlockNotifications{}
+		flt  = neorpc.NotificationFilter{}
+	)
+	if filter != nil {
+		flt = *filter.Copy()
+	}
+	if err := c.performRequest("getblocknotifications", []any{blockHash.StringLE(), flt}, resp); err != nil {
```
Comment (AnnaShaleva) [inline pkg/rpcclient/rpc.go:985]: This way you're going to have empty notification filter passed to RPC server in case of nil `filter`. And what we need in this case is just a blockhash. See more carefully how `ReceiveExecutionNotifications` works and do the same for this handler.

File: pkg/rpcclient/rpc_test.go [hunk 4]
```diff
@@ -1859,6 +1859,12 @@ var rpcClientErrorCases = map[string][]rpcClientErrorCase{
 				return nil, c.ValidateAddress("")
 			},
 		},
+		{
+			name: "getblocknotifications_unmarshalling_error",
+			invoke: func(c *Client) (any, error) {
+				return c.GetBlockNotifications(util.Uint256{}, nil)
```
Comment (AnnaShaleva) [inline pkg/rpcclient/rpc_test.go:1865]: I also need a good test scenario, I haven't seen it in the list of tests.

File: pkg/rpcclient/rpc.go (hunk 3 above)
Comment (AnnaShaleva) [reply pkg/rpcclient/rpc.go:985]: Add an end-to-end test for this particular case to `pkg/services/rpcsrv/client_test.go` to ensure that nil filter will be handled properly by server.

File: pkg/services/rpcsrv/client_test.go [hunk 5]
```diff
@@ -2488,3 +2488,39 @@ func TestClient_NEP24(t *testing.T) {
...
+		})
+		require.NoError(t, err)
+		require.Empty(t, bn)
+	})
+	t.Run("good", func(t *testing.T) {
+		name := "Transfer"
+		bn, err := c.GetBlockNotifications(chain.GetHeaderHash(1), &neorpc.NotificationFilter{
+			Name: &name,
+		})
+		fmt.Println(bn)
```
Comment (AnnaShaleva) [inline pkg/services/rpcsrv/client_test.go:2522]: Remove it.

File: pkg/rpcclient/rpc.go (hunk 3 above)
Comment (AnnaShaleva) [reply pkg/rpcclient/rpc.go:985]: Not fixed.
```
params := []any{blockHash.StringLE()}
if flt != nil {
	params = append(params, *flt)
}
if err := c.performRequest("getblocknotifications", params, resp); err != nil {
...
```
And after a bit of thinking, I see that filter copy here may be omitted because filter won't likely be reused by the caller's code.

**Review by AnnaShaleva — APPROVED** [review]
